IPv6 Support for Internal APIM Node Communication

### Description

The following errors are observed in an IPv6-only cluster when communication occurs between the Gateway and Control Plane, preventing JMS synchronization between the nodes.

```
GW

TID: [] [] [2026-04-23 09:04:37,548] ERROR {org.wso2.carbon.apimgt.common.jms.JMSTaskManager} - Error creating JMS consumer for Siddhi-JMS-Consumer javax.jms.JMSException: Error registering consumer: org.wso2.andes.AMQChannelClosedException: Error: Queue name: tmp_[fd00_0_0_0_0_0_0_3]_55068_10 cannot contain any of following symbols ~!@#;%^*()+={}|<>"', and space. " / " can only use as the tenant separator
 [error code 541: internal error]
	at org.wso2.andes.client.AMQSession$6.execute(AMQSession.java:2187)
	at org.wso2.andes.client.AMQSession$6.execute(AMQSession.java:2130)
	at org.wso2.andes.client.AMQConnectionDelegate_8_0.executeRetrySupport(AMQConnectionDelegate_8_0.java:340)
	at org.wso2.andes.client.AMQConnection$3.run(AMQConnection.java:665)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:569)
	at org.wso2.andes.client.AMQConnection.executeRetrySupport(AMQConnection.java:662)
	at org.wso2.andes.client.failover.FailoverRetrySupport.execute(FailoverRetrySupport.java:102)
	at org.wso2.andes.client.AMQSession.createConsumerImpl(AMQSession.java:2195)
	at org.wso2.andes.client.AMQSession.createConsumer(AMQSession.java:1100)
	at org.wso2.carbon.apimgt.common.jms.utils.JMSUtils.createConsumer(JMSUtils.java:501)
	at org.wso2.carbon.apimgt.common.jms.JMSTaskManager$MessageListenerTask.createConsumer(JMSTaskManager.java:1010)
	at org.wso2.carbon.apimgt.common.jms.JMSTaskManager$MessageListenerTask.getMessageConsumer(JMSTaskManager.java:865)
	at org.wso2.carbon.apimgt.common.jms.JMSTaskManager$MessageListenerTask.receiveMessage(JMSTaskManager.java:612)
	at org.wso2.carbon.apimgt.common.jms.JMSTaskManager$MessageListenerTask.run(JMSTaskManager.java:533)
	at org.apache.axis2.transport.base.threads.NativeWorkerPool$1.run(NativeWorkerPool.java:172)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: org.wso2.andes.AMQChannelClosedException: Error: Queue name: tmp_[fd00_0_0_0_0_0_0_3]_55068_10 cannot contain any of following symbols ~!@#;%^*()+={}|<>"', and space. " / " can only use as the tenant separator
 [error code 541: internal error]
	at org.wso2.andes.client.handler.ChannelCloseMethodHandler.methodReceived(ChannelCloseMethodHandler.java:100)
	at org.wso2.andes.client.handler.ClientMethodDispatcherImpl.dispatchChannelClose(ClientMethodDispatcherImpl.java:163)
	at org.wso2.andes.framing.amqp_0_91.ChannelCloseBodyImpl.execute(ChannelCloseBodyImpl.java:140)
	at org.wso2.andes.client.state.AMQStateManager.methodReceived(AMQStateManager.java:111)
	at org.wso2.andes.client.protocol.AMQProtocolHandler.methodBodyReceived(AMQProtocolHandler.java:554)
	at org.wso2.andes.client.protocol.AMQProtocolSession.methodFrameReceived(AMQProtocolSession.java:467)
	at org.wso2.andes.framing.AMQMethodBodyImpl.handle(AMQMethodBodyImpl.java:92)
	at org.wso2.andes.client.protocol.AMQProtocolHandler$2.run(AMQProtocolHandler.java:487)
	at org.wso2.andes.pool.Job.processAll(Job.java:111)
	at org.wso2.andes.pool.Job.run(Job.java:158)
	... 3 more



CP

TID: [] [] [2026-04-23 09:02:02,826] ERROR {org.wso2.andes.server.protocol.AMQProtocolEngine} - Closing channel (19) due to internal Error org.wso2.andes.AMQChannelException: Queue name: tmp_[fd00_0_0_0_0_0_0_3]_55068_10 cannot contain any of following symbols ~!@#;%^*()+={}|<>"', and space. " / " can only use as the tenant separator
 [error code 541: internal error]
	at org.wso2.andes.framing.AMQMethodBodyImpl.getChannelException(AMQMethodBodyImpl.java:73)
	at org.wso2.andes.server.handler.QueueDeclareHandler.methodReceived(QueueDeclareHandler.java:193)
	at org.wso2.andes.server.handler.ServerMethodDispatcherImpl.dispatchQueueDeclare(ServerMethodDispatcherImpl.java:600)
	at org.wso2.andes.framing.amqp_0_91.QueueDeclareBodyImpl.execute(QueueDeclareBodyImpl.java:187)
	at org.wso2.andes.server.state.AMQStateManager.methodReceived(AMQStateManager.java:170)
	at org.wso2.andes.server.protocol.AMQProtocolEngine.methodFrameReceived(AMQProtocolEngine.java:389)
	at org.wso2.andes.framing.AMQMethodBodyImpl.handle(AMQMethodBodyImpl.java:92)
	at org.wso2.andes.server.protocol.AMQProtocolEngine.frameReceived(AMQProtocolEngine.java:334)
	at org.wso2.andes.server.protocol.AMQProtocolEngine.dataBlockReceived(AMQProtocolEngine.java:283)
	at org.wso2.andes.server.protocol.AMQProtocolEngine$1.run(AMQProtocolEngine.java:252)
	at org.wso2.andes.pool.Job.processAll(Job.java:111)
	at org.wso2.andes.pool.Job.run(Job.java:158)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```




### Steps to Reproduce

Prerequisites

Docker installed with Rancher Desktop
WSO2 APIM 4.2.0 Docker image available locally as wso2am:4.2.0.patch
Patched startup script at ~/api-manager-ipv6.sh (see Step 1)
CP config at ~/cp-conf/repository/conf/deployment.toml (see Step 2)
GW config at ~/gw-conf/repository/conf/deployment.toml (see Step 3)

## Step 1 — Patch the startup script (one-time)

```bash
docker run --rm --name apim-tmp -d wso2am:4.2.0.patch
docker cp apim-tmp:/home/wso2carbon/wso2am-4.2.0/bin/api-manager.sh ~/api-manager-ipv6.sh
docker rm -f apim-tmp

# macOS
sed -i '' 's/-Djava.net.preferIPv4Stack=true/-Djava.net.preferIPv6Stack=true/' ~/api-manager-ipv6.sh

chmod +x ~/api-manager-ipv6.sh

# Verify
grep "preferIPv" ~/api-manager-ipv6.sh
# Expected: -Djava.net.preferIPv6Stack=true
```

---

## Step 2 — Create the CP deployment.toml (one-time)

```bash
mkdir -p ~/cp-conf/repository/conf

cat > ~/cp-conf/repository/conf/deployment.toml << 'EOF'
[server]
hostname = "localhost"
offset = 1
node_ip = "$env{NODE_IP}"
server_role = "control-plane"

[user_store]
type = "database_unique_id"

[super_admin]
username = "admin"
password = "admin"
create_admin_account = true

[database.shared_db]
type = "h2"
url = "jdbc:h2:./repository/database/WSO2SHARED_DB;DB_CLOSE_ON_EXIT=FALSE"
username = "wso2carbon"
password = "wso2carbon"

[keystore.tls]
file_name = "wso2carbon.jks"
type = "JKS"
password = "wso2carbon"
alias = "wso2carbon"
key_password = "wso2carbon"

[truststore]
file_name = "client-truststore.jks"
type = "JKS"
password = "wso2carbon"

# [transport.http]
# properties.port = 9764
# properties.proxyPort = 80

# [transport.https]
# properties.port = 9444
# properties.proxyPort = 443

[[apim.gateway.environment]]
name = "Default"
type = "hybrid"
provider = "wso2"
display_in_api_console = true
description = "This is a hybrid gateway that handles both production and sandbox token traffic."
show_as_token_endpoint_url = true
service_url = "https://apim-gw-ipv6:9444/services/"
username = "${admin.username}"
password = "${admin.password}"
ws_endpoint = "ws://apim-gw-ipv6:9099"
wss_endpoint = "wss://apim-gw-ipv6:8099"
http_endpoint = "http://apim-gw-ipv6:8281"
https_endpoint = "https://apim-gw-ipv6:8244"
websub_event_receiver_http_endpoint = "http://apim-gw-ipv6:9021"
websub_event_receiver_https_endpoint = "https://apim-gw-ipv6:8021"

[[event_listener]]
id = "token_revocation"
type = "org.wso2.carbon.identity.core.handler.AbstractIdentityHandler"
name = "org.wso2.is.notification.ApimOauthEventInterceptor"
order = 1

[event_listener.properties]
notification_endpoint = "https://localhost:9444/internal/data/v1/notify"
username = "${admin.username}"
password = "${admin.password}"
'header.X-WSO2-KEY-MANAGER' = "default"

[apim.key_manager]
service_url = "https://localhost:9444/services/"
username = "$ref{super_admin.username}"
password = "$ref{super_admin.password}"

[apim.event_hub]
enable = true
username = "$ref{super_admin.username}"
password = "$ref{super_admin.password}"
service_url = "https://localhost:9444/services/"
event_listening_endpoints = ["tcp://localhost:5673"]

[apim.devportal]
url = "https://localhost:9444/devportal"

[apim.analytics]
enable = false

[transport.passthru_https.sender.parameters]
HostnameVerifier = "AllowAll"
EOF
```

---

## Step 3 — Create the GW deployment.toml (one-time)

```bash
mkdir -p ~/gw-conf/repository/conf

cat > ~/gw-conf/repository/conf/deployment.toml << 'EOF'
[server]
hostname = "localhost"
offset = 1
node_ip = "$env{NODE_IP}"
server_role = "gateway-worker"

[user_store]
type = "database_unique_id"

[super_admin]
username = "admin"
password = "admin"
create_admin_account = true

[database.shared_db]
type = "h2"
url = "jdbc:h2:./repository/database/WSO2SHARED_DB;DB_CLOSE_ON_EXIT=FALSE"
username = "wso2carbon"
password = "wso2carbon"

[keystore.tls]
file_name = "wso2carbon.jks"
type = "JKS"
password = "wso2carbon"
alias = "wso2carbon"
key_password = "wso2carbon"

[truststore]
file_name = "client-truststore.jks"
type = "JKS"
password = "wso2carbon"

# [transport.http]
# properties.port = 9764
# properties.proxyPort = 80

# [transport.https]
# properties.port = 9444
# properties.proxyPort = 443

[apim.key_manager]
service_url = "https://wso2am-pattern-3-am-cp-service:9444/services/"
username = "$ref{super_admin.username}"
password = "$ref{super_admin.password}"

[apim.sync_runtime_artifacts.gateway]
gateway_labels = ["Default"]

[apim.throttling]
username = "$ref{super_admin.username}"
password = "$ref{super_admin.password}"
service_url = "https://wso2am-pattern-3-am-cp-service:9444/services/"
throttle_decision_endpoints = ["tcp://wso2am-pattern-3-am-cp-1-service:5673",
                                "tcp://wso2am-pattern-3-am-cp-2-service:5673"]
enable_unlimited_tier = true
enable_header_based_throttling = false
enable_jwt_claim_based_throttling = false
enable_query_param_based_throttling = false

[[apim.throttling.url_group]]
traffic_manager_urls = ["tcp://wso2am-pattern-3-am-cp-1-service:9612"]
traffic_manager_auth_urls = ["ssl://wso2am-pattern-3-am-cp-1-service:9712"]
type = "loadbalance"

[[apim.throttling.url_group]]
traffic_manager_urls = ["tcp://wso2am-pattern-3-am-cp-2-service:9612"]
traffic_manager_auth_urls = ["ssl://wso2am-pattern-3-am-cp-2-service:9712"]
type = "loadbalance"

[apim.analytics]
enable = false

[transport.passthru_https.sender.parameters]
HostnameVerifier = "AllowAll"
EOF
```

---

## Step 4 — Create the Docker network

```bash
docker network inspect ipv6-only-test > /dev/null 2>&1 || \
  docker network create --ipv6 --subnet "fd00::/64" ipv6-only-test
```

Verify:

```bash
docker network inspect ipv6-only-test \
  --format '{{range .IPAM.Config}}Subnet: {{.Subnet}}{{"\n"}}{{end}}'
```

Expected:
```
Subnet: fd00::/64
Subnet: 172.x.x.x/16
```

---

## Step 5 — Remove stale containers

```bash
docker rm -f apim-cp-ipv6 apim-gw-ipv6 2>/dev/null; true
```

---

## Step 6 — Start the Control Plane


```bash
docker run -d --name apim-cp-ipv6 \
  --network ipv6-only-test \
  --ip6 fd00::2 \
  -e PROFILE_NAME=control-plane \
  -e NODE_IP=fd00::2 \
  -p 9444:9444 \
  --add-host "apim-gw-ipv6:fd00::3" \
  -v ~/api-manager-ipv6.sh:/home/wso2carbon/wso2am-4.2.0/bin/api-manager.sh \
  -v ~/cp-conf/repository/conf/deployment.toml:/home/wso2carbon/wso2-config-volume/repository/conf/deployment.toml \
  wso2am:4.2.0.patch
```

Wait until Andes is ready (~2 min):

---

## Step 7 — Start the Gateway


```bash
docker run -d --name apim-gw-ipv6 \
  --network ipv6-only-test \
  --ip6 fd00::3 \
  -e PROFILE_NAME=gateway-worker \
  -e NODE_IP=fd00::3 \
  -p 8281:8281 \
  -p 8244:8244 \
  --add-host "wso2am-pattern-3-am-cp-service:fd00::2" \
  --add-host "wso2am-pattern-3-am-cp-1-service:fd00::2" \
  --add-host "wso2am-pattern-3-am-cp-2-service:fd00::2" \
  -v ~/api-manager-ipv6.sh:/home/wso2carbon/wso2am-4.2.0/bin/api-manager.sh \
  -v ~/gw-conf/repository/conf/deployment.toml:/home/wso2carbon/wso2-config-volume/repository/conf/deployment.toml \
  wso2am:4.2.0.patch
```

---

## Step 8 — Observe the error

**Gateway logs** :

```bash
docker logs -f apim-gw-ipv6
```

**CP logs**:

```bash
docker logs -f apim-cp-ipv6 
```






### Version

4.2

### Environment Details (with versions)

_No response_

--- a/modules/andes-core/client/src/main/java/org/wso2/andes/client/protocol/AMQProtocolSession.java
+++ b/modules/andes-core/client/src/main/java/org/wso2/andes/client/protocol/AMQProtocolSession.java
@@ -391,8 +391,8 @@ public class AMQProtocolSession implements AMQVersionAwareProtocolSession
         {
             id = _queueId++;
         }
-        // convert '.', '/', ':' and ';' to single '_', for spec compliance and readability
-        String localAddress = _protocolHandler.getLocalAddress().toString().replaceAll("[./:;]", "_");
+        // convert '.', '/', ':', ';', '[' and ']' to single '_', for spec compliance and readability
+        String localAddress = _protocolHandler.getLocalAddress().toString().replaceAll("[./:;\\[\\]]", "_");
         String queueName = "tmp_" + localAddress + "_" + id;
         return new AMQShortString(queueName.replaceAll("_+", "_"));
     }

--- a/modules/andes-core/client/src/main/java/org/wso2/andes/client/protocol/AMQProtocolSession.java
+++ b/modules/andes-core/client/src/main/java/org/wso2/andes/client/protocol/AMQProtocolSession.java
@@ -391,8 +391,8 @@ public class AMQProtocolSession implements AMQVersionAwareProtocolSession
         {
             id = _queueId++;
         }
-        // convert '.', '/', ':', ';', '[' and ']' to single '_', for spec compliance and readability
-        String localAddress = _protocolHandler.getLocalAddress().toString().replaceAll("[./:;\\[\\]]", "_");
+        // convert '.', '/', ':', ';', '[', ']' and '%' to single '_', for spec compliance and readability
+        String localAddress = _protocolHandler.getLocalAddress().toString().replaceAll("[./:;\\[\\]%]", "_");
         String queueName = "tmp_" + localAddress + "_" + id;
         return new AMQShortString(queueName.replaceAll("_+", "_"));
     }

